common/flatpak-prune.c

```diff
@@ -0,0 +1,813 @@
+                    int            depth,
+                    gboolean       dry_run,
+                    int           *out_objects_total,
+                    int           *out_objects_pruned,
+                    guint64       *out_pruned_object_size_total,
+                    GCancellable  *cancellable,
+                    GError       **error)
+{
+  g_autoptr(FlatpakOstreeObjectNameBag) reachable = object_name_bag_new ();
+  OtPruneData data = { 0, };
+  g_autoptr(GTimer) timer = NULL;
+
+  /* This version only handles archive repos, if called for something else call ostree */
+  if (ostree_repo_get_mode (repo) != OSTREE_REPO_MODE_ARCHIVE)
+    {
+      OstreeRepoPruneFlags flags = OSTREE_REPO_PRUNE_FLAGS_REFS_ONLY;
+      if (dry_run)
+        flags |= OSTREE_REPO_PRUNE_FLAGS_NO_PRUNE;
+
+      return ostree_repo_prune (repo, flags, depth,
+                                out_objects_total, out_objects_pruned, out_pruned_object_size_total,
+                                cancellable, error);
+    }
+
+  {
+    /* shared lock in this region, see locking strategy above */
+    glnx_autofd int lock_fd = -1;
+
+    if (!get_repo_lock (repo, LOCK_SH, &lock_fd, cancellable, error))
+      return FALSE;
+
+    timer = g_timer_new ();
+    g_debug ("Finding reachable objects, unlocked (depth=%d)", depth);
+    g_timer_start (timer);
+
+    g_usleep (1000*1000*10);
+
+    if (!traverse_reachable_refs_unlocked (repo, depth, reachable, cancellable, error))
+      return FALSE;
+
+    g_timer_stop (timer);
+    g_debug ("Elapsed time: %.1f sec",  g_timer_elapsed (timer, NULL));
+  }
+
+  {
+    /* exclusive lock in this region, see locking strategy above */
+    glnx_autofd int lock_fd = -1;
+
+    if (!get_repo_lock (repo, LOCK_EX, &lock_fd, cancellable, error))
+      return FALSE;
+
+    timer = g_timer_new ();
+    g_debug ("Finding reachable objects, locked (depth=%d)", depth);
+    g_timer_start (timer);
+
+    if (!traverse_reachable_refs_unlocked (repo, depth, reachable, cancellable, error))
+      return FALSE;
+
+    data.repo = repo;
+    data.reachable = reachable;
+    data.dont_prune = dry_run;
+
+    g_timer_stop (timer);
+    g_debug ("Elapsed time: %.1f sec",  g_timer_elapsed (timer, NULL));
+
+    g_debug ("Pruning unreachable objects");
+    g_timer_start (timer);
+
+    if (!prune_unreachable_loose_objects (repo, &data, cancellable, error))
+      return FALSE;
+
+    g_timer_stop (timer);
+    g_debug ("Elapsed time: %.1f sec",  g_timer_elapsed (timer, NULL));
+  }
+
+  /* Prune static deltas outside lock to avoid conflict with its exclusive lock */
+  if (!dry_run)
+    {
+      g_debug ("Pruning static deltas");
+      g_timer_start (timer);
+
+      if (!ostree_repo_prune_static_deltas (repo, NULL, cancellable, error))
+        return FALSE;
+
+      g_timer_stop (timer);
+      g_debug ("Elapsed time: %.1f sec",  g_timer_elapsed (timer, NULL));
+    }
+
+  *out_objects_total = data.n_reachable + data.n_unreachable;
```

Comment:
Would be nice if the fix to this line were in the commit that introduced it rather than an unrelated commit

Reply:
Yeah, I just noticed it there. I think I will merge all the prune.c related changes into one commit in the end, but keeping it separate now for easier review.